File: packages/clinical-contracts/src/rule4/severity/assertionBinding.ts
import { RULE4_PHASE_TARGET_ROLE_VALUES, type Rule4PhaseTargetRole } from '../phase/types.js';
import type { Rule4FormulaSeverityRecord, Rule4SeverityEvidenceAssertion } from './types.js';

const FLARE: Rule4PhaseTargetRole = 'CURRENT_ACUTE_FLARE';
const CHRONIC: Rule4PhaseTargetRole = 'UNDERLYING_CHRONIC_TARGET';

export type CrossRoleGuardOptions = {
  label: 'SYNTHETIC' | 'PRODUCTION';
  trustedSyntheticBindingBypass: boolean;
};

export function syntheticBindingBypassActive(options: CrossRoleGuardOptions): boolean {
  return options.label === 'SYNTHETIC' && options.trustedSyntheticBindingBypass === true;
}

export function isKnownTargetRole(value: unknown): value is Rule4PhaseTargetRole {
  return (
    typeof value === 'string' &&
    (RULE4_PHASE_TARGET_ROLE_VALUES as readonly string[]).includes(value)
  );
}

export function validateAssertionBoundTargetRoles(
  records: readonly Rule4FormulaSeverityRecord[],
): void {
  for (const record of records) {
    for (const assertion of record.severityEvidenceAssertions) {
      const raw = assertion.boundTargetRole;
      if (raw != null && !isKnownTargetRole(raw)) {
        throw new Error('RULE4_SEVERITY_BOUND_TARGET_ROLE_INVALID');
      }
    }
  }
}

export type ResolvedAssertionBinding = {
  evidenceItemId: string;
  receivingFormulaSlotId: string;
  receivingFormulaTargetId: string;
  receivingTargetRole: Rule4PhaseTargetRole;
  boundFormulaSlotId: string;
  boundFormulaTargetId: string;
  boundTargetRole: Rule4PhaseTargetRole;
  bindingAuthoritative: boolean;
  missingAcuteChronicAuthority: boolean;
};

export function resolveAssertionBinding(
  assertion: Rule4SeverityEvidenceAssertion,
  record: Rule4FormulaSeverityRecord,
  options: CrossRoleGuardOptions,
): ResolvedAssertionBinding {
  const receivingTargetRole = record.targetRole;
  const explicitRole = assertion.boundTargetRole;

  let boundTargetRole: Rule4PhaseTargetRole;
  let missingAcuteChronicAuthority = false;

  if (explicitRole != null) {
    boundTargetRole = explicitRole;
  } else if (receivingTargetRole === FLARE || receivingTargetRole === CHRONIC) {
    if (syntheticBindingBypassActive(options)) {
      boundTargetRole = receivingTargetRole;
    } else {
      missingAcuteChronicAuthority = true;
      boundTargetRole = receivingTargetRole;
    }
  } else {
    boundTargetRole = receivingTargetRole;
  }

  const boundFormulaSlotId = assertion.boundFormulaSlotId ?? record.formulaSlotId;
  const boundFormulaTargetId = assertion.boundFormulaTargetId ?? record.formulaTargetId;

  const bindingAuthoritative =
    !missingAcuteChronicAuthority &&
    (explicitRole != null ||
      (receivingTargetRole === 'STANDARD_FORMULA_TARGET' &&
        boundFormulaSlotId === record.formulaSlotId &&
        boundFormulaTargetId === record.formulaTargetId) ||
      (syntheticBindingBypassActive(options) &&
        (receivingTargetRole === FLARE || receivingTargetRole === CHRONIC)));

  return {
    evidenceItemId: assertion.evidenceItemId,
    receivingFormulaSlotId: record.formulaSlotId,
    receivingFormulaTargetId: record.formulaTargetId,
    receivingTargetRole,
    boundFormulaSlotId,
    boundFormulaTargetId,
    boundTargetRole,
    bindingAuthoritative,
    missingAcuteChronicAuthority,
  };
}

export function isFlareOrChronic(role: Rule4PhaseTargetRole): boolean {
  return role === FLARE || role === CHRONIC;
}

export function isAcuteChronicCross(
  boundRole: Rule4PhaseTargetRole,
  receivingRole: Rule4PhaseTargetRole,
): boolean {
  return (
    (boundRole === FLARE && receivingRole === CHRONIC) ||
    (boundRole === CHRONIC && receivingRole === FLARE)
  );
}

File: apps/clinical-engine/src/ehas2_clinical_engine/rule4/severity/assertion_binding.py
from __future__ import annotations

FLARE = "CURRENT_ACUTE_FLARE"
CHRONIC = "UNDERLYING_CHRONIC_TARGET"
STANDARD = "STANDARD_FORMULA_TARGET"

KNOWN_TARGET_ROLES = {STANDARD, FLARE, CHRONIC}


def synthetic_binding_bypass_active(options: dict) -> bool:
    return options.get("label") == "SYNTHETIC" and options.get("trusted_synthetic_binding_bypass") is True


def is_known_target_role(value: object) -> bool:
    return isinstance(value, str) and value in KNOWN_TARGET_ROLES


def validate_assertion_bound_target_roles(records: list[dict]) -> None:
    for record in records:
        for assertion in record.get("severity_evidence_assertions") or []:
            raw = assertion.get("bound_target_role")
            if raw is not None and not is_known_target_role(raw):
                raise ValueError("RULE4_SEVERITY_BOUND_TARGET_ROLE_INVALID")


def resolve_assertion_binding(assertion: dict, record: dict, options: dict) -> dict:
    receiving_target_role = record["target_role"]
    explicit_role = assertion.get("bound_target_role")

    missing_acute_chronic_authority = False
    if explicit_role is not None:
        bound_target_role = explicit_role
    elif receiving_target_role in {FLARE, CHRONIC}:
        if synthetic_binding_bypass_active(options):
            bound_target_role = receiving_target_role
        else:
            missing_acute_chronic_authority = True
            bound_target_role = receiving_target_role
    else:
        bound_target_role = receiving_target_role

    bound_formula_slot_id = assertion.get("bound_formula_slot_id") or record["formula_slot_id"]
    bound_formula_target_id = assertion.get("bound_formula_target_id") or record["formula_target_id"]

    binding_authoritative = not missing_acute_chronic_authority and (
        explicit_role is not None
        or (
            receiving_target_role == STANDARD
            and bound_formula_slot_id == record["formula_slot_id"]
            and bound_formula_target_id == record["formula_target_id"]
        )
        or (
            synthetic_binding_bypass_active(options)
            and receiving_target_role in {FLARE, CHRONIC}
        )
    )

    return {
        "evidence_item_id": assertion["evidence_item_id"],
        "receiving_formula_slot_id": record["formula_slot_id"],
        "receiving_formula_target_id": record["formula_target_id"],
        "receiving_target_role": receiving_target_role,
        "bound_formula_slot_id": bound_formula_slot_id,
        "bound_formula_target_id": bound_formula_target_id,
        "bound_target_role": bound_target_role,
        "binding_authoritative": binding_authoritative,
        "missing_acute_chronic_authority": missing_acute_chronic_authority,
    }


def is_flare_or_chronic(role: str) -> bool:
    return role in {FLARE, CHRONIC}


def is_acute_chronic_cross(bound_role: str, receiving_role: str) -> bool:
    return (bound_role == FLARE and receiving_role == CHRONIC) or (
        bound_role == CHRONIC and receiving_role == FLARE
    )
